openedx/core/djangoapps/credit/verification_access.py

```diff
@@ -0,0 +1,202 @@
+"""
+Create in-course reverification access groups in a course.
+
+We model the rules as a set of user partitions, one for each
+verification checkpoint in a course.
+
+For example, suppose that a course has two verification checkpoints,
+one at midterm A and one at the midterm B.
+
+Then the user partitions would look like this:
+
+Midterm A:  |-- ALLOW --|-- DENY --|
+Midterm B:  |-- ALLOW --|-- DENY --|
+
+where the groups are defined as:
+
+* ALLOW: The user has access to content gated by the checkpoint.
+* DENY: The user does not have access to content gated by the checkpoint.
+
+"""
+import logging
+
+from util.db import generate_int_id
+from openedx.core.djangoapps.credit.utils import get_course_blocks
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.partitions.partitions import Group, UserPartition
+
+
+log = logging.getLogger(__name__)
+
+
+VERIFICATION_SCHEME_NAME = "verification"
+VERIFICATION_BLOCK_CATEGORY = "edx-reverification-block"
+
+
+def create_verification_partitions(course_key):
+    """
+    Create a user partition for each verification checkpoint in the course.
+
+    This will modify the published version of the course descriptor.
+    It ensures that any in-course reverification XBlocks in the course
+    have an associated user partition.  Other user partitions (e.g. cohorts)
+    will be preserved.  Partitions associated with deleted reverification checkpoints
+    will be marked as inactive and will not be used to restrict access.
+
+    Arguments:
+        course_key (CourseKey): identifier for the course.
+
+    Returns:
+        None
+    """
+    # Batch all the queries we're about to do and suppress
+    # the "publish" signal to avoid an infinite call loop.
+    with modulestore().bulk_operations(course_key, emit_signals=False):
+
+        # Retrieve all in-course reverification blocks in the course
+        icrv_blocks = get_course_blocks(course_key, VERIFICATION_BLOCK_CATEGORY)
+
+        # Update the verification definitions in the course descriptor
+        # This will also clean out old verification partitions if checkpoints
+        # have been deleted.
+        _set_verification_partitions(course_key, icrv_blocks)
+
+
+def _unique_partition_id(course):
```

Comment:
I put it here because it's importing a function from the `utils` Djangoapp, and I wanted to avoid contaminating `xmodule/partitions` with Django dependencies.  Am I overthinking this?